stratumv2/src/message/types/bytes.rs

```diff
@@ -0,0 +1,243 @@
+use crate::error::{Error, Result};
+use crate::message::parse::{ByteParser, Deserializable, Serializable};
+use crate::message::types::fixed::U24;
+use std::io;
+
+/// An internal macro that implements a B0 type that is restricted according to a MAX_LENGTH.
+macro_rules! impl_sized_B0 {
+    ($type:ident, $length_type:ident) => {
+        impl_sized_B0!($type, $length_type, $length_type::MAX as usize);
+    };
+    ($type:ident, $length_type:ident, $max_length:expr) => {
+        #[derive(Debug, Clone, PartialEq)]
+        pub struct $type {
+            pub(crate) length: $length_type,
+            pub(crate) data: Vec<u8>,
+        }
+
+        impl $type {
+            const MAX_LENGTH: usize = $max_length;
+
+            /// The constructor enforces the Vec<u8> input size as the MAX_LENGTH. A
+            /// RequirementError will be returned if the input byte size is greater than the
+            /// MAX_LENGTH.
+            pub fn new<T: Into<Vec<u8>>>(value: T) -> Result<$type> {
+                let value = value.into();
+                if value.len() > Self::MAX_LENGTH {
+                    return Err(Error::RequirementError(
+                        "bytes size cannot be greater than MAX_LENGTH".into(),
+                    ));
+                }
+
+                use std::convert::TryInto;
+                Ok($type {
+                    length: value.len().try_into().unwrap(),
+                    data: value,
+                })
+            }
+        }
+
+        /// PartialEq implementation allowing direct comparison between the B0 type and Vec<u8>.
+        impl PartialEq<Vec<u8>> for $type {
+            fn eq(&self, other: &Vec<u8>) -> bool {
+                self.data == *other
+            }
+        }
+
+        /// PartialEq implementation allowing direct comparison between Vec<u8> and the B0 type.
+        impl PartialEq<$type> for Vec<u8> {
+            fn eq(&self, other: &$type) -> bool {
+                *self == other.data
+            }
+        }
+
+        /// From trait implementation that allows a B0 to be converted into a Vec<u8>.
+        impl From<$type> for Vec<u8> {
+            fn from(b: $type) -> Self {
+                b.data
+            }
+        }
+
+        /// Deserialize trait implementation that allows a B0 to be deserialized from a ByteParser.
+        impl Deserializable for $type {
+            fn deserialize(parser: &mut ByteParser) -> Result<$type> {
+                // Parse the length header before the buffer.
+                let header_length = $length_type::deserialize(parser)?;
+                // Then parse the byte buffer.
+                let bytes = parser.next_by(header_length.clone().into())?;
+
+                $type::new(bytes.iter().cloned().collect::<Vec<u8>>())
```

Comment:
I think simply doing `$type::new(bytes)` is sufficient, since argument in the constructor is `Into<Vec<u8>>`, so the constructor will call `into()` and transform the byte slice to a `Vec<u8>`

Reply:
Done.